### PR TITLE
Address PROTOTYPE comments around ParenthesizedOperation and not eras…

### PR DESCRIPTION
**Pull request by 333fred**
…ing UnaryOperations uncecessarily.

@dotnet/roslyn-compiler @AlekseyTs @dotnet/analyzer-ioperation for review.

### CODE REVIEW

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >while (condition.Kind == OperationKind.Parenthesized) [](start = 12, length = 53)

It feels like this should be done inside the loop too, after each time we dig through a logical Not. Consider changing the loop below to:

while(true)
{
    Dig through parentesized
    Break if not logical not
    Dig through logical Not
}

Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: Please add relevant test.

---
In reply to: [193943036](https://github.com/dotnet/roslyn/pull/27611#discussion_r193943036) [](ancestors = 193943036)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >if (unOp.Operand.Kind != OperationKind.UnaryOperator && [](start = 16, length = 55)

Consider if the following approach is simpler and involves less logic duplication:
- Remember the last Not that we about to dig through here.
- pass it to the ProduceIsSense
- Change ProduceIsSense to simply update the operand of that logical Not if we have it and need to invert the value.

BTW, it looks like ProduceIsSense can be inlined into this function because there are no more consumers.

File: src/Compilers/VisualBasic/Test/IOperation/IOperation/IOperationTests_IBinaryOperatorExpression.vb
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/IOperation/IOperation/IOperationTests_IBinaryOperatorExpression.vb]: >a = b AndAlso (Not c) [](start = 8, length = 21)

This scenario looks sufficiently good to demonstrate the reuse of existing Not operator, but I think it doesn't quite show the impact from skipping parenthesized.
Try adding tests for ```a = b OrElse (Not c andalso Not d)``` and ```a = b OrElse Not (c orElse d)```. The should produce graphs equivalent to ```a = b OrElse Not c andalso Not d```, modulo the syntax associated with nodes.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 1)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (333fred) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: @AlekseyTs this was the bug. Rather than stopSense (which is true in the case of something like `!(b || c)`), we need to be looking at the actual desired value at the end.

**Comment by 333fred** [comment]
@AlekseyTs addressed feedback.

---
In reply to: [395639442](https://github.com/dotnet/roslyn/pull/27611#issuecomment-395639442) [](ancestors = 395639442)

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: >// revert constant value if we have one. [](start = 75, length = 40)

Please open an issue for this comment.

**Review by AlekseyTs — APPROVED** [review]
LGTM (iteration 2)

**Comment by AlekseyTs** [comment]
@333fred, @mavasani, @cston, @dotnet/roslyn-compiler, @dotnet/analyzer-ioperation Please review

**Review by mavasani — APPROVED** [review]

**Comment by AlekseyTs** [comment]
@cston, @dotnet/roslyn-compiler Please review, need second review.

**Review by cston — APPROVED** [review]

**Comment by AlekseyTs** [comment]
Test windows_debug_vs-integration_prtest please

**Pull request merged**
